File: handy-python/quiz/q18_kth_largest_element_in_an_array_medium.py
"""
Q18) Kth Largest Element in an Array

AI-BEST:
- Quickselect with in-place partitioning around a pivot.
- Time: O(n) average, O(n^2) worst case, Space: O(1) extra

AI-EASY:
- Sort descending and index directly.
- Time: O(n log n), Space: depends on sort implementation
"""


def kth_largest_best(nums: list[int], k: int) -> int:
    values = nums[:]
    target = len(values) - k
    left = 0
    right = len(values) - 1

    def partition(lo: int, hi: int, pivot_index: int) -> int:
        pivot_value = values[pivot_index]
        values[pivot_index], values[hi] = values[hi], values[pivot_index]
        store_index = lo
        for i in range(lo, hi):
            if values[i] < pivot_value:
                values[store_index], values[i] = values[i], values[store_index]
                store_index += 1
        values[store_index], values[hi] = values[hi], values[store_index]
        return store_index

    # Step 1: Partition around a pivot so smaller values land on the left.
    # Why: Quickselect needs a partially ordered array to focus the search.
    while left <= right:
        pivot_index = (left + right) // 2
        pivot_index = partition(left, right, pivot_index)

        # Step 2: Narrow the search to the side that contains the target index.
        # Why: Only one side can contain the kth largest after partitioning.
        if pivot_index == target:
            return values[pivot_index]
        if pivot_index < target:
            left = pivot_index + 1
        else:
            right = pivot_index - 1

    # Step 3: Return the target value once the pivot lands on it.
    # Why: The loop converges when the correct order statistic is isolated.
    return values[target]


def kth_largest_easy(nums: list[int], k: int) -> int:
    # Step 1: Sort the numbers in descending order.
    # Why: The kth largest is then just a direct lookup.
    ordered = sorted(nums, reverse=True)

    # Step 2: Pick the element at position k - 1.
    # Why: Descending order makes the target index easy to compute.
    return ordered[k - 1]


def run_demo() -> None:
    print("Q18: Kth Largest Element in an Array")
    samples = [
        ([3, 2, 1, 5, 6, 4], 2),
        ([3, 2, 3, 1, 2, 4, 5, 5, 6], 4),
        ([1], 1),
    ]
    for nums, k in samples:
        print(f"Input: nums={nums}, k={k}")
        print("  BEST:", kth_largest_best(nums, k))
        print("  EASY:", kth_largest_easy(nums, k))


if __name__ == "__main__":
    run_demo()

File: handy-ts/quiz/q18_kth_largest_element_in_an_array_medium.ts
/**
 * Q18) Kth Largest Element in an Array
 *
 * AI-BEST:
 * Quickselect partitions the array until the target index lands in place.
 * Time: O(n) average, O(n^2) worst case, Space: O(1) extra
 *
 * AI-EASY:
 * Sort descending and take the kth item.
 * Time: O(n log n), Space: depends on sort implementation
 */

function kthLargestBest(nums: number[], k: number): number {
  const arr = [...nums];
  const targetIndex = arr.length - k;
  let left = 0;
  let right = arr.length - 1;

  // Step 1: Partition around a pivot because quickselect only needs the side that contains the answer.
  while (left <= right) {
    const pivotIndex = partition(arr, left, right);

    // Step 2: Narrow to the side containing the target because the other side cannot contain the kth largest value.
    if (pivotIndex === targetIndex) {
      // Step 3: Stop immediately once the pivot lands on the target because that value is the answer.
      return arr[pivotIndex];
    }
    if (pivotIndex < targetIndex) {
      left = pivotIndex + 1;
    } else {
      right = pivotIndex - 1;
    }
  }

  return -1;
}

function partition(arr: number[], left: number, right: number): number {
  const pivotValue = arr[right];
  let storeIndex = left;

  for (let i = left; i < right; i++) {
    if (arr[i] <= pivotValue) {
      [arr[storeIndex], arr[i]] = [arr[i], arr[storeIndex]];
      storeIndex++;
    }
  }

  [arr[storeIndex], arr[right]] = [arr[right], arr[storeIndex]];
  return storeIndex;
}

function kthLargestEasy(nums: number[], k: number): number {
  // Step 1: Sort descending because the kth largest item then sits at index k - 1.
  const sorted = [...nums].sort((a, b) => b - a);

  // Step 2: Pick the target index because the order has already been fully established.
  const result = sorted[k - 1];

  // Step 3: Return the selected value because it is the requested kth largest element.
  return result;
}

function runDemo(): void {
  console.log("Q18: Kth Largest Element in an Array");
  const cases = [
    { nums: [3, 2, 1, 5, 6, 4], k: 2 },
    { nums: [3, 2, 3, 1, 2, 4, 5, 5, 6], k: 4 },
    { nums: [1], k: 1 },
  ];

  for (const { nums, k } of cases) {
    console.log(`Input: nums=${JSON.stringify(nums)}, k=${k}`);
    console.log("  BEST:", kthLargestBest(nums, k));
    console.log("  EASY:", kthLargestEasy(nums, k));
  }
}

runDemo();

export {};
